datafusion/core/src/execution/context/mod.rs

```diff
@@ -1246,7 +1273,42 @@ impl QueryPlanner for DefaultQueryPlanner {
             .await
     }
 }
+/// Crates and registers a function from [CreateFunction] statement
+///
+/// It is intended to handle `CREATE FUNCTION` statements
+/// and interact with [SessionState] to registers new udfs.
+///
+/// Datafusion `SQL` dialect does not support `CREATE FUNCTION`
+/// in generic dialect, so dialect should be changed to `PostgreSQL`
+///
+/// ```rust, no_run
+/// # use datafusion::execution::config::SessionConfig;
+/// SessionConfig::new().set_str("datafusion.sql_parser.dialect", "PostgreSQL");
+/// ```
+#[async_trait]
+pub trait FunctionFactory: Sync + Send {
+    // TODO: I don't like having RwLock Leaking here, who ever implements it
+    //       has to depend ot `parking_lot`. I'f we expose &mut SessionState it
+    //       may keep lock of too long.
+    //
+    //       Not sure if there is better approach.
+    //
+
+    /// Handles creation of user defined function specified in [CreateFunction] statement
+    async fn create(
+        &self,
+        state: Arc<RwLock<SessionState>>,
```

Comment:
It seems like `create` here handles two things:
1. Creation of the function
2. Registering the function with the SessionState

What if we made `create` only responsible for creating a function and made the caller responsible for registering it? I think that would make the API simpler / not involve  RW lock

For example, something like
```rust
/// A function defined by a FunctionFactory
enum NewFunction  {
  Scalar(ScalarUDF),
  Aggregate(AggregateUDF),
  Window(WindowUDF)
}

pub trait FunctionFactory: Sync + Send {
...
async fn create(
        &self,
        statement: CreateFunction,
) -> Result<NewFunction>;
...
}
```




Another alternative might be rather than passing the `SessionState`, maybe we could pass a reference to `&mut dyn FunctionRegistry` which would also avoid the RwLock 🤔

Reply:
Should we want to implement sql user defined functions, we would need extension to cache logical expressions and provide them to logical analyser so it can replace udf call with the expression.

Id argue we need state or configuration as a parameter as well. Let me do other changes, as they are straightforward and we can try to find most ergonomic function

Reply:
I definitely like this approach, but it has lock problem if we want to provide a `config` parameter. 

Why would we need `config`? `FunctionFactory` may store configuration for model repository url, or similar

Reply:
I think you are right. I mentioned `config` because I think there was a comment about it before.

Reply:
Sorry for confusion @alamb, I believe i fix it with right amount of .clone() on right places :).
Please have a look now, I believe it looks all right. I still do not like `remove(...)` could be implemented with another enum ... have o think about it